### ISSUE TITLE
Add aarch64 neon simd support

### ISSUE THREAD

**Issue by Count-Count** [BurntSushi/memchr#76]
Currently there are only an x86_64-specific implementations using SIMD instructions.

As described in #75 I want to try to port the implementation to aarch64 using the code in [/src/x86/sse42.rs](../blob/master/src/x86/sse42.rs) as a template and the NEON intrinsics from [core::arch::aarch64](https://doc.rust-lang.org/core/arch/aarch64/index.html) (currently nightly only).

**Comment by BurntSushi** [comment]
Thanks for opening an issue!

I'm somewhat surprised to see that you want to use sse42 as a template. Did you mean sse2? I ask because sse42 uses a fairly specialized intrinsic provided by Intel. I'm not as familiar with Neon, but I'd be surprised if such a thing existed there.

Otherwise, my main concern with this is the fact that Neon intrinsics requires nightly. In general, I don't really want my crates to have code that requires nightly, even if it's an opt-in feature. In the past, I've tried this (a few times), and it almost always results in a much higher maintenance burden than I want.

I'm not quite sure how to mitigate this unfortunately. :-/

**Comment by Count-Count** [comment]
You are totally right, I meant sse2.rs, but linked the wrong file. It is still not a straight-forward translation as aarch64 neon does not have anything like _mm_movemask_epi8 afaik.

Do you think that core::arch::aarch64 will become stable eventually or will there be major changes? If no major changes are expected I would be OK with trying to implement it now and keeping it in my fork until it becomes stable.

**Comment by BurntSushi** [comment]
Not sure. My guess is that there won't be any major changes since the goal is really just to expose what the platform provides. So there isn't a ton of design work there AIUI. But I don't think the Neon functions have a champion, so it's kind of stuck.

I do somewhat suspect that the new Apple hardware will cause a lot more people to care about this though, so perhaps there is hope for stabilization sometime soon.

**Comment by bertjwregeer** [comment]
I am guessing this support will be necessary before this will compile on Rust on aarch64 on macOS Big Sur on Apple Silicon M1.

**Comment by BurntSushi** [comment]
Why would it be necessary?

**Comment by bertjwregeer** [comment]
I was attempting to cargo install ripgrep, and it failed to build memchr. I came looking for an issue that mentioned aarch64.

**Comment by BurntSushi** [comment]
Please file a bug. This issue should not be necessary for compiling this crate. Please include as many details as possible.

**Comment by bertjwregeer** [comment]
@BurntSushi ignore my comments. I just found: https://github.com/shepmaster/rust/blob/silicon/silicon/README.md which mentions there may be issues due to some parts being x86_64 and some parts being compiled arm64.

Using `RUSTC=$(rustup which rustc) $(rustup which cargo) install --verbose ripgrep`

Worked, and memchr was compiled as well. I can delete my comments.

**Comment by BurntSushi** [comment]
Please don't delete comments. There's no point. And if someone else runs into the same problem, maybe a search will turn up this conversation.

**Comment by BurntSushi** [comment]
Update: I think the best path forward here will be to use `std::simd`, whenever that becomes stable. `std::simd` is meant to be a portable SIMD abstraction. I _believe_ memchr's use of SIMD is simple enough where it will fit easily on top of a portable API. We'll just need to benchmark it carefully to make sure we get the same codegen that we get now.

**Comment by BurntSushi** [comment]
According to [this comment on irlo](https://internals.rust-lang.org/t/using-a-more-efficient-string-matching-algorithm/16719/32?u=burntsushi), aarch64 doesn't have a "move mask" operation, which is central to how the current memchr and memmem implementations work. I spent ten minutes looking at the [unstable `std::simd` API](https://doc.rust-lang.org/std/simd/index.html), but couldn't figure out whether it has an equivalent portable movemask operation. (If aarch64 really doesn't have one, then I wouldn't expect the portable API to have it either, unless they can somehow emulate it efficiently. But still, it's worth checking.) So if we really can't use movemask as is from x86, then we'll need to figure something else out from aarch64.

I personally know nothing about what kinds of SIMD routines are available on aarch64. So we'll need to either look at how other folks do this (I'd look at Go's standard library or any heavily optimized libc implementation) for aarch64. _Or_ someone who has expertise in aarch64 SIMD.

**Comment by redzic** [comment]
I'm currently working on implementing memchr for aarch64 NEON for this crate (although there is already https://github.com/ARM-software/optimized-routines which has memchr for NEON and even SVE, but it doesn't implement it with unrolling which might improve performance even more).

It is true that aarch64 doesn't have a movemask instruction. However, it does have an instruction that adds adjacent pairwise elements, and because of this, there is a really interesting trick that you can use to get a 64 bit integer that you can then count the trailing zeros on to find the matching index.

For the sake of simplicity, I'll pretend that NEON SIMD lanes are only 2 bytes wide. Basically what you can do is use `cmeq` (which basically functions exactly the same as SSE2's `cmpeq`), then `and` this with `0xF00F`.

Now the first byte, if it matched, will have the value `0xF0`, and otherwise it will just be 0. The second byte, if it matched, will have the value `0x0F`. Normally it's just `0xFF` for both bytes, but it's different since we `and` the result with `0xF00F`. Now, if we add these two elements together (which can be done efficiently in SIMD with NEON's `addp` instruction), we get the combined mask of these two elements. On a full 128-bit wide register, you can do this to get a 64-bit mask, count the trailing zeros, then divide the result by 4 (since the mask has 4 bits set for each element) to get the first matching index.

So far, I've written an extremely basic implementation of memchr for aarch64 based on this, which currently is effectively just a simplified version of the optimized-routines memchr (only works with lengths that are multiples of 16, fixed needle of 5 (although this is a really easy fix, but this is just an example anyway), and no unrolling, and no special handling to ensure accesses are aligned):

```assembly
; x0 - pointer to src u8 array
; x1 - length of array (bytes)
function memchr, export=1
        mov             x3, x1
        mov	        w2, 0xf00f
        ; v2 = mask register
	dup	        v2.8h, w2

        movi.16b        v1, #5
L(loop):
        ldr             q0, [x0], #16
        cmeq            v0.16b, v0.16b, v1.16b
        umaxp           v3.16b, v0.16b, v0.16b
        fmov            x2, d3
        cbnz            x2, L(match_found)
        subs            x1, x1, #16
        b.ne            L(loop)
        ; no match was found
        mov             x0, -1
        ret
L(match_found):
        sub             x3, x3, x1
        and             v0.16b, v0.16b, v2.16b
        addp            v2.16b, v0.16b, v0.16b
        fmov            x0, d2
        rbit            x0, x0
        clz             x0, x0
        lsr             x0, x0, 2
        add             x0, x3, x0
        ret
endfunc
```

Some of the aarch64 NEON intrinsics seem to be stable as of Rust 1.59, but I'm not sure if all of the instructions needed for memchr are covered yet. Not sure if's acceptable to use assembly directly for this crate if it's the case that the intrinsics are not available on stable yet.

**Comment by BurntSushi** [comment]
@redzic That's awesome! Thank you for working on this. :-)

> Some of the aarch64 NEON intrinsics seem to be stable as of Rust 1.59, but I'm not sure if all of the instructions needed for memchr are covered yet. Not sure if's acceptable to use assembly directly for this crate if it's the case that the intrinsics are not available on stable yet.

So generally speaking, it's probably a little worse than this. I generally try to keep the MSRV for memchr somewhat conservative, so even "it's stable" is usually not enough. By "conservative," I usually like to make sure it builds on Rust compilers going back 1 year.

With that said, for high value adds like this, I can be amenable to version sniffing `rustc` in `build.rs`, and then enabling code paths automatically if the Rust version is high enough. So it seems likely that we'll want to go that route.

In principle I'm not opposed to Assembly, particularly given that we now have inline Assembly, but I do worry about my ability to maintain such a thing. I'd really really prefer to stick to intrinsics. And I think it would be reasonable to use this crate as a motivating reason to push stabilizing of necessary intrinsics if they aren't already stable.

**Comment by redzic** [comment]
Yeah that makes sense, I definitely agree that intrinsics are significantly easier to maintain than Assembly code. I only suggested it because that's probably the only good way to immediately get NEON support out there, even for very old rustc versions.

I had a quick look in the Rust stdlib, and it looks like the intrinsics for `cmeq` (there seem to be many, apparently for different data types) are all nightly-only at the moment.

Given that, perhaps the best route (at least for now, assuming we don't want to mess with the unstable nightly intrinsics) is to implement it via inline assembly? Although I think the *best* solution would be to get the intrinsics stabilized so we can just use them directly (although of course the `build.rs` stuff would still be necessary for some time because of the MSRV related issues you mentioned).

**Comment by BurntSushi** [comment]
To be honest, I think I'd need to see it and try to understand what kind of, if any, build issues arise as a result of using it. At the very least, if it causes issues, I may need to wind up backing it out. One of the main problems here is that I have no arm hardware at all _and_ also no experience with it.

There are other issues too, like, does arm have the equivalent of runtime CPU feature detection that we'll need to do before using certain instructions?

I think my bottom line is that I'm not 100% opposed to using inline assembly for this, but I'm pretty close to it. The only real reason I'm open to it is that I haven't tried it before, and therefore, don't have any concrete experience to fall back on.

**Comment by BurntSushi** [comment]
Getting those intrinsics stabilized is definitely my preferred route.

**Comment by BurntSushi** [comment]
The other thing here is the portable SIMD API, aka, `std::simd`. My original plan was to move to that and hope that everything works out and we get great codegen on lots of targets. But maybe that's wishful thinking. (And `std::simd` isn't stable yet anyway.)

Have you seen the `Vector` trait in the `memmem` module? Ideally we could bring that to `memchr` too, which is what #103 does. And then just impl that trait and add the necessary glue to make neon work. But that really requires the intrinsics, or at least, Assembly functions that are equivalent to the intrinsics.

Maybe that might be a good alternative? Use intrinsics, and for any that aren't stable, polyfill them with inline asm. But otherwise don't just go full hog into inline asm.

**Comment by redzic** [comment]
> Have you seen the Vector trait in the memmem module? Ideally we could bring that to memchr too, which is what https://github.com/BurntSushi/memchr/pull/103 does. And then just impl that trait and add the necessary glue to make neon work. But that really requires the intrinsics, or at least, Assembly functions that are equivalent to the intrinsics.

I have indeed seen the trait. While it seems like it would work great for x86, I think something like this would need to be done so that the trait can be generically reused to get an efficient NEON implementation:

1. Add a distinct function solely for checking if the cmpeq return vector is all zeroes. The reason is that on x86, this would be done by manually calling `movemask`, but this is an extremely suboptimal way to check for the loop exit condition on ARM, so this needs to be abstracted away so that the more efficient implementation is used on ARM, and `pmovmskb` is still used on x86 for this. 
2. Remove `movemask` and replace it with a function that returns the first matching index. This function would assume that the input vector was returned from `cmpeq` and isn't just any arbitrary vector, and also that there is a matching value, the idea being that you're supposed to call the function outlined in 1. before checking the matching index. The reason this is needed is that it would be very slow to directly implement an exact equivalent of `movemask` on NEON, but it's possible to do something that's still fast that fulfills (one of) its intended purposes, which is to calculate the matching index.

> Maybe that might be a good alternative? Use intrinsics, and for any that aren't stable, polyfill them with inline asm. But otherwise don't just go full hog into inline asm.

Yeah definitely. I probably should have clarified that by "use inline assembly", I didn't mean to directly use it in the actual memchr implementation, but rather use inline assembly to implement the remaining unstable intrinsics so that at least the code still looks like intrinsics instead of inline assembly. I only mentioned inline assembly for this reason, sorry if it seemed like my intention was to use them directly for memchr.

**Comment by redzic** [comment]
Hmm, strangely enough, even though the docs say they are a nightly-only API, the supposedly unstable intrinsics compile completely fine on stable. Huh, I guess this will be easier than I thought as far as stabilizing intrinsics go.

**Comment by BurntSushi** [comment]
Interesting. Can you point to at least one intrinsic that claims itself as unstable but compiles on stable?

**Comment by BurntSushi** [comment]
> Yeah definitely. I probably should have clarified that by "use inline assembly", I didn't mean to directly use it in the actual memchr implementation, but rather use inline assembly to implement the remaining unstable intrinsics (in this crate only) so that at least the code still looks like intrinsics instead of inline assembly. I only mentioned inline assembly for this reason, sorry if it seemed like my intention was to use them directly for memchr. Obviously the best route is still to get the intrinsics stabilized.

Ah okay, good. I feel a lot better about inline asm (if it's necessary) in this context then.

**Comment by redzic** [comment]
> Interesting. Can you point to at least one intrinsic that claims itself as unstable but compiles on stable?

Yes. For reference, [`vld1q_u8`](https://doc.rust-lang.org/core/arch/aarch64/fn.vld1q_u8.html) has no disclaimer in the docs that it's nightly-only.

There are others, but [`vceqq_u8`](https://doc.rust-lang.org/core/arch/aarch64/fn.vceqq_u8.html), [`vget_low_u64`](https://doc.rust-lang.org/core/arch/aarch64/fn.vget_low_u64.html), and [`vdupq_n_u8`](https://doc.rust-lang.org/core/arch/aarch64/fn.vdupq_n_u8.html) all have a disclaimer saying "This is a nightly-only experimental API. (stdsimd [#48556](https://github.com/rust-lang/rust/issues/48556))"

However, despite these disclaimers, they still compile on stable: https://godbolt.org/z/he4TsWaxs (I can confirm locally also that it compiles, not just on godbolt)

**Comment by BurntSushi** [comment]
Ah, so thankfully, all of those are actually stable. It looks like the stability attribute is applied only on aarch64 targets, and it looks like rustdoc doesn't quite know how to deal with that. You can see the "stable" attribute here: https://doc.rust-lang.org/src/core/up/up/stdarch/crates/core_arch/src/arm_shared/neon/generated.rs.html#864

This one is applied unconditionally and rustdoc displays it correctly as "since 1.59.0": https://doc.rust-lang.org/src/core/up/up/stdarch/crates/core_arch/src/aarch64/neon/mod.rs.html#565

The other two are also stable, but it is only applied on aarch64. So the same rustdoc bug.

Next time I'm on a keyboard I'll file a rustdoc bug if one hasn't been created yet.

TL;DR: all of those are thankfully truly stable since 1.59.0, but there is a rustdoc bug preventing it from showing up in the docs in some cases.

**Comment by redzic** [comment]
I've written the initial implementation with 4x unrolling and handling main loop so that the accesses are aligned. The structure is basically the same as the SSE2 implementation. These are the results on an 8-core M1 Pro compared to the fallback implementation (sorry that this comment is so long, I couldn't figure out how to easily condense all the benchmarks):

```
memchr1/krate/huge/never                                                                            
                        time:   [6.9165 us 6.9268 us 6.9378 us]
                        thrpt:  [79.863 GiB/s 79.990 GiB/s 80.109 GiB/s]
                 change:
                        time:   [-65.976% -65.804% -65.612%] (p = 0.00 < 0.05)
                        thrpt:  [+190.80% +192.43% +193.91%]
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) low mild
  1 (10.00%) high severe

memchr1/krate/huge/rare time:   [9.0733 us 9.0985 us 9.1171 us]                                    
                        thrpt:  [60.773 GiB/s 60.898 GiB/s 61.067 GiB/s]
                 change:
                        time:   [-61.140% -60.975% -60.808%] (p = 0.00 < 0.05)
                        thrpt:  [+155.15% +156.25% +157.33%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild

memchr1/krate/huge/uncommon                                                                           
                        time:   [106.83 us 109.84 us 111.71 us]
                        thrpt:  [4.9599 GiB/s 5.0445 GiB/s 5.1866 GiB/s]
                 change:
                        time:   [-19.469% -17.426% -15.101%] (p = 0.00 < 0.05)
                        thrpt:  [+17.787% +21.103% +24.175%]
                        Performance has improved.

memchr1/krate/huge/common                                                                            
                        time:   [396.45 us 397.64 us 398.72 us]
                        thrpt:  [1.3896 GiB/s 1.3934 GiB/s 1.3976 GiB/s]
                 change:
                        time:   [-33.615% -33.208% -32.797%] (p = 0.00 < 0.05)
                        thrpt:  [+48.802% +49.718% +50.636%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild

Benchmarking memchr1/krate/huge/verycommon: Collecting 10 samples in estimated 2.0070 s (3135 iteration                                                                                                       memchr1/krate/huge/verycommon                        
                        time:   [639.08 us 639.86 us 640.91 us]
                        thrpt:  [885.26 MiB/s 886.72 MiB/s 887.80 MiB/s]
                 change:
                        time:   [-31.868% -31.319% -30.762%] (p = 0.00 < 0.05)
                        thrpt:  [+44.430% +45.600% +46.775%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

memchr1/krate/small/never                                                                            
                        time:   [9.6147 ns 9.6200 ns 9.6327 ns]
                        thrpt:  [64.198 GiB/s 64.283 GiB/s 64.318 GiB/s]
                 change:
                        time:   [-61.420% -61.260% -61.109%] (p = 0.00 < 0.05)
                        thrpt:  [+157.13% +158.13% +159.20%]
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  2 (20.00%) high severe

memchr1/krate/small/rare                                                                            
                        time:   [12.429 ns 12.445 ns 12.467 ns]
                        thrpt:  [49.605 GiB/s 49.690 GiB/s 49.755 GiB/s]
                 change:
                        time:   [-58.856% -58.661% -58.455%] (p = 0.00 < 0.05)
                        thrpt:  [+140.70% +141.90% +143.05%]
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe

memchr1/krate/small/uncommon                                                                           
                        time:   [41.929 ns 41.941 ns 41.964 ns]
                        thrpt:  [14.736 GiB/s 14.745 GiB/s 14.749 GiB/s]
                 change:
                        time:   [-26.905% -26.566% -26.163%] (p = 0.00 < 0.05)
                        thrpt:  [+35.434% +36.176% +36.809%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

memchr1/krate/small/common                                                                            
                        time:   [302.78 ns 303.03 ns 303.46 ns]
                        thrpt:  [2.0378 GiB/s 2.0407 GiB/s 2.0424 GiB/s]
                 change:
                        time:   [+30.595% +31.316% +31.900%] (p = 0.00 < 0.05)
                        thrpt:  [-24.185% -23.848% -23.427%]
                        Performance has regressed.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

Benchmarking memchr1/krate/small/verycommon: Collecting 10 samples in estimated 2.0000 s (3.0M iteratio                                                                                                       memchr1/krate/small/verycommon                        
                        time:   [666.38 ns 667.31 ns 668.44 ns]
                        thrpt:  [947.34 MiB/s 948.94 MiB/s 950.26 MiB/s]
                 change:
                        time:   [+53.506% +54.781% +55.822%] (p = 0.00 < 0.05)
                        thrpt:  [-35.824% -35.393% -34.856%]
                        Performance has regressed.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe

memchr1/krate/tiny/never                                                                            
                        time:   [4.3417 ns 4.3436 ns 4.3485 ns]
                        thrpt:  [14.778 GiB/s 14.794 GiB/s 14.801 GiB/s]
                 change:
                        time:   [-7.2656% -6.7773% -6.3445%] (p = 0.00 < 0.05)
                        thrpt:  [+6.7743% +7.2700% +7.8349%]
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe

memchr1/krate/tiny/rare time:   [7.1398 ns 7.1501 ns 7.1604 ns]                                    
                        thrpt:  [8.9745 GiB/s 8.9874 GiB/s 9.0005 GiB/s]
                 change:
                        time:   [-32.480% -32.232% -31.990%] (p = 0.00 < 0.05)
                        thrpt:  [+47.037% +47.562% +48.104%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild

memchr1/krate/tiny/uncommon                                                                            
                        time:   [16.105 ns 16.111 ns 16.122 ns]
                        thrpt:  [3.9860 GiB/s 3.9887 GiB/s 3.9902 GiB/s]
                 change:
                        time:   [-47.333% -47.067% -46.799%] (p = 0.00 < 0.05)
                        thrpt:  [+87.967% +88.919% +89.873%]
                        Performance has improved.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

memchr1/krate/tiny/common                                                                           
                        time:   [54.887 ns 54.920 ns 54.978 ns]
                        thrpt:  [1.1689 GiB/s 1.1701 GiB/s 1.1708 GiB/s]
                 change:
                        time:   [+19.091% +19.698% +20.454%] (p = 0.00 < 0.05)
                        thrpt:  [-16.981% -16.457% -16.031%]
                        Performance has regressed.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

memchr1/krate/empty/never                                                                           
                        time:   [155.04 ps 155.31 ps 155.56 ps]
                        thrpt:  [0.0000   B/s 0.0000   B/s 0.0000   B/s]
                 change:
                        time:   [-0.6220% -0.0503% +0.4559%] (p = 0.88 > 0.05)
                        thrpt:  [-0.4538% +0.0504% +0.6259%]
                        No change in performance detected.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe
```

Basically seems like there are performance improvements across the board, except for some regressions with small haystacks. Maybe some further tuning could be done for the small haystacks, I'm open to ideas here.

**Comment by redzic** [comment]
I also wanted to post the actual code here for some visibility.

```rust
pub fn memchr(n1: u8, haystack: &[u8]) -> Option<usize> {
    use std::arch::aarch64::*;
    use std::mem::transmute;

    const VEC_SIZE: usize = 16;
    const VEC_SIZE_4X: usize = VEC_SIZE * 4;

    const MASK4: u64 = generate_mask64();

    /// Returns true if the all bits in the regsiter are set to 0.
    #[inline(always)]
    unsafe fn eq0(x: uint8x16_t) -> bool {
        transmute::<_, u64>(vget_low_u64(transmute(vpmaxq_u8(x, x)))) == 0
    }

    unsafe {
        if haystack.is_empty() {
            return None;
        }

        let start_ptr = haystack.as_ptr();
        let end_ptr = start_ptr.add(haystack.len());

        if haystack.len() < VEC_SIZE {
            let mut ptr = start_ptr;
            while ptr != end_ptr {
                if *ptr == n1 {
                    return Some(ptr as usize - start_ptr as usize);
                }

                ptr = ptr.add(1);
            }
            return None;
        }

        let repmask1 = vdupq_n_u16(0xF00F);
        let repmask4 = vdupq_n_u64(MASK4);

        let n1v = vdupq_n_u8(n1);

        // search first 16 bytes (unaligned)
        let x1 = vld1q_u8(start_ptr);
        let cmpmask = vceqq_u8(x1, n1v);

        if !eq0(cmpmask) {
            let cmpmask = vandq_u8(cmpmask, transmute(repmask1));
            let combined = vpaddq_u8(transmute(cmpmask), transmute(cmpmask));
            let comb_low: u64 = transmute(vget_low_u64(transmute(combined)));

            return Some(comb_low.trailing_zeros() as usize / 4);
        }

        // get next aligned pointer
        let mut aligned_ptr = ((VEC_SIZE * ((start_ptr as usize) / VEC_SIZE))
            + VEC_SIZE) as *const u8;

        // aligned unroll 4x end pointer
        let a4x_end_ptr = aligned_ptr
            .add(VEC_SIZE_4X * ((haystack.len() - VEC_SIZE) / VEC_SIZE_4X));

        while aligned_ptr != a4x_end_ptr {
            let x1 = vld1q_u8(aligned_ptr.add(0 * VEC_SIZE));
            let x2 = vld1q_u8(aligned_ptr.add(1 * VEC_SIZE));
            let x3 = vld1q_u8(aligned_ptr.add(2 * VEC_SIZE));
            let x4 = vld1q_u8(aligned_ptr.add(3 * VEC_SIZE));

            let cmp1 = vceqq_u8(x1, n1v);
            let cmp2 = vceqq_u8(x2, n1v);
            let cmp3 = vceqq_u8(x3, n1v);
            let cmp4 = vceqq_u8(x4, n1v);

            let cmp = vorrq_u8(vorrq_u8(cmp1, cmp2), vorrq_u8(cmp3, cmp4));

            if !eq0(cmp) {
                let cmp1 = vandq_u8(transmute(repmask4), cmp1);
                let cmp2 = vandq_u8(transmute(repmask4), cmp2);
                let cmp3 = vandq_u8(transmute(repmask4), cmp3);
                let cmp4 = vandq_u8(transmute(repmask4), cmp4);

                let reduce1 = vpaddq_u8(cmp1, cmp2);
                let reduce2 = vpaddq_u8(cmp3, cmp4);
                let reduce3 = vpaddq_u8(reduce1, reduce2);
                let reduce4 = vpaddq_u8(reduce3, reduce3);

                let low64: u64 = transmute(vget_low_u64(transmute(reduce4)));

                let offset = aligned_ptr as usize - start_ptr as usize;

                return Some(offset + low64.trailing_zeros() as usize);
            }

            aligned_ptr = aligned_ptr.add(VEC_SIZE * 4);
        }

        let remaining_vec_len =
            VEC_SIZE * ((end_ptr as usize - a4x_end_ptr as usize) / VEC_SIZE);

        if remaining_vec_len != 0 {
            let a1x_end_ptr = a4x_end_ptr.add(remaining_vec_len);

            while aligned_ptr != a1x_end_ptr {
                let x1 = vld1q_u8(aligned_ptr);
                let cmp = vceqq_u8(x1, n1v);

                if !eq0(cmp) {
                    let cmp = vandq_u8(transmute(repmask1), cmp);
                    let combined = vpaddq_u8(cmp, cmp);
                    let comb_low: u64 =
                        transmute(vget_low_u64(transmute(combined)));

                    let offset = aligned_ptr as usize - start_ptr as usize;

                    return Some(
                        offset + comb_low.trailing_zeros() as usize / 4,
                    );
                }

                aligned_ptr = aligned_ptr.add(VEC_SIZE);
            }
        }

        if end_ptr as isize - aligned_ptr as isize != 0 {
            // search last 16 bytes (unaligned)

            let x1 = vld1q_u8(end_ptr.sub(VEC_SIZE));
            let cmp = vceqq_u8(x1, n1v);

            if !eq0(cmp) {
                let cmp = vandq_u8(transmute(repmask1), cmp);
                let combined = vpaddq_u8(cmp, cmp);
                let comb_low: u64 =
                    transmute(vget_low_u64(transmute(combined)));

                let offset =
                    end_ptr.sub(VEC_SIZE) as usize - start_ptr as usize;

                return Some(offset + comb_low.trailing_zeros() as usize / 4);
            }
        }

        None
    }
}
```

I think from here memchr2, memchr3, and memrchr, etc. need to also be implemented.

**Comment by thomcc** [comment]
Regarding stability (since it was referenced in the other thread), the MSRV of this doesn't quite matter if the build.rs has checks like `target_has_feature` on aarch64 <https://github.com/BurntSushi/memchr/blob/master/build.rs#L33>.

While it will not allow for runtime detection, `neon` is always present on most aarch64 targets anyway (on aarch64-unknown-linux-gnu, aarch64-apple-darwin, etc), and the target-feature stabilized shortly after core::arch::aarch64, so if `neon` is provided in `CARGO_CFG_TARGET_FEATURES`.

This won't help for 32 bit arm, but that isn't stable yet anyway.

**Comment by BurntSushi** [comment]
Yeah, basically, when it comes to aarch64 I basically know nothing. For example, in order to use AVX2 on x86, it is very very clear that you need runtime feature detection or else it's probably not going to get used much. I just naturally assumed the same would be true for aarch64. But yeah, there's no reason that my assumption is true.

Thanks for pointing this out. @redzic do you agree with @thomcc that `neon` is ubiquitous enough that we don't actually need runtime detection? I'll note that this isn't necessarily just about ISA popularity. For example, I would still use runtime feature detection for SSSE3, even though it's probably pretty rare these days to come across a x86 chip without SSSE3. The issue is that 1) many x86 targets don't include SSSE3 by default and 2) there are still some weird environments, like VMs, where the CPU features available are a restricted set. (Sorry, I don't know much more about (2), but I know bug reports cropped up about them back when I used to ship ripgrep binaries specifically compiled for SSSE3.)

**Comment by thomcc** [comment]
Notably runtime detection returns `cfg!(target_feature = ...)` a lot of the time at the moment anyway. The only targets that currently implement runtime detection in the `stdarch/std_detect` code are `aarch64-unknown-freebsd` and `aarch64-unknown-linux-*` (because this can be done without an OS, and while it's not particularly difficult to do so with an OS, the way the code is set up now it would be undesirable).

Notably, all of `rustc --print cfg --target aarch64-unknown-freebsd`, `rustc --print cfg --target aarch64-unknown-linux-gnu`, and `rustc --print cfg --target aarch64-unknown-freebsd`

It's a bit like sse2, with the caveat that it's only architecturally guaranteed on ARMv8-A (A = Application) and not on ARMv8-R (R = Realtime) or ARMv8-M (M = 'Mbedded). The latter two are not clearly ever going to support runtime detection, in std_detect though, since those would always be `no_std` targets.

**Comment by CeleritasCelery** [comment]
I tested @redzic implementation and it seems to work on Aarch64. Is not having good neon feature detection the only thing blocking this from getting merged?

**Comment by MathGeniusJodie** [comment]
some ideas
https://community.arm.com/arm-community-blogs/b/infrastructure-solutions-blog/posts/porting-x86-vector-bitmask-optimizations-to-arm-neon

**Comment by BurntSushi** [comment]
Closed by #129
